docs/tutorial_completers.rst

```diff
@@ -0,0 +1,130 @@
+should return ``None`` or an empty set, which will cause xonsh to move on and
+try to use the next completer.
+
+Occasionally, completers will need to return a match that does not actually
+start with ``prefix``.  In this case, a completer should instead return a tuple
+``(completions, prefixlength)``, where ``completions`` is the set of
+appropriate completions, and ``prefixlength`` is the number of characters in
+``line`` that should be treated as part of the completion.
+
+The docstring of a completer should contain a brief description of its
+functionality, which will be displayed by ``list-completers``.
+
+Three examples follow.  For more examples, see the source code of the completers
+xonsh actually uses, in the ``xonsh.completers`` module.
+
+.. code-block:: python
+
+    def dummy_completer(prefix, line, begidx, endidx, ctx):
+        '''
+        Completes everything with options "lou" and "carcolh",
+        regardless of the value of prefix.
+        '''
+        return {"lou", "carcolh"}
+    
+    def python_context_completer(prefix, line, begidx, endidx, ctx):
+        '''
+        Completes based on the names in the current Python environment
+        '''
+        return {i for i in ctx if i.startswith(prefix)}
+
+    def unbeliever_completer(prefix, line, begidx, endidx, ctx):
+        '''
+        Replaces "lou carcolh" with "snail" if tab is pressed after typing
+        "lou" and when typing "carcolh"
+        '''
+        if 'carcolh'.startswith(prefix) and line[:begidx].split()[-1] == 'lou':
+            return ({'snail'}, len('lou ') + len(prefix))
+
+
+Registering a Completer
+=======================
+
+Once you have created a completion function, you can add it to the list of
+active completers via the ``register-completer`` command::
+
+    Usage:
+        register-completer NAME FUNC [POS]
+
+``NAME`` is a unique name to use in the listing
+
+``FUNC`` is the name of a completer function to use.
+
+``POS`` (optional) is a position into the list of completers at which the new completer should be added.  It can be one of the following values:
+
+* ``"start"`` indicates that the completer should be added to the start of the list of completers (it should be run before all others)
+* ``"end"`` indicates that the completer should be added to the end of the list of completers (it should be run after all others)
+* ``">KEY"``, where ``KEY`` is a pre-existing name, indicates that this should be added after the completer named ``KEY``
+* ``"<KEY"``, where ``KEY`` is a pre-existing name, indicates that this should be added before the completer named ``KEY``
+
+.. note:: It is also possible to manipulate ``__xonsh_completers__`` directly,
+          but this is the preferred method.
+
+Removing a Completer
+====================
+
+To remove a completer from the list of active completers, run
+``remove-completer NAME``, where ``NAME`` is the unique identifier associated
```

Comment:
I guess I would prefer an interface that was `completer add` and `completer rm` or similar than two different commands.  I'd be happy to modify this later, though, and should not be blocking here, if you are amicable.